tests/test_util.py

```diff
@@ -286,3 +244,105 @@ def perturb_function(f, var, k, delta):
   f_value = f()
   var.assign(old_value)
   return f_value
+
+
+def _five_point_stencil(f, var, i, delta):
+  """Computes the five point stencil.
+
+  See wikipedia page on "five point stencil",
+  https://en.wikipedia.org/wiki/Five-point_stencil
+  Note: the error of this method scales with delta ** 4.
+
+  Args:
+    f: Callable taking no arguments and returning a possibly nested structure
+      whose atomic elements are `tf.Tensor`.
+    var: `tf.Variable` in which to differentiate `f`.
+    i: Entry of `var` to perturb.
+    delta: Size of the fundamental perturbation in the stencil.
+
+  Returns:
+    stencil: The five point stencil approximation of the jacobian of `f()` with
+      respect to entry `i` of `var`.
+  """
+  forward_twice = perturb_function(f, var, i, 2.0 * delta)
+  forward_once = perturb_function(f, var, i, delta)
+  backward_once = perturb_function(f, var, i, -1.0 * delta)
+  backward_twice = perturb_function(f, var, i, -2.0 * delta)
+  numerator = tf.nest.map_structure(
+      lambda a, b, c, d: -1.0 * a + 8.0 * b - 8.0 * c + d, forward_twice,
+      forward_once, backward_once, backward_twice)
+  stencil = tf.nest.map_structure(lambda x: x / (12.0 * delta), numerator)
+  return stencil
+
+
+def approximate_gradient(f, variables, delta=1e-1):
+  """Approximates the gradient of f using five point stencil.
+
+  Suppose the input function returns a possibly nested structure `r` under
+  gradient tape `t`.  Then this function returns an approximation to
+  `t.gradient(r, variables, unconnected_gradients=tf.UnconnectedGradients.ZERO)`
+
+  Args:
+    f: Callable taking no arguments and returning a possibly nested structure
+      whose atomic elements are `tf.Tensor`.
+    variables: Possibly nested structure of `tf.Variable` in which to
+      differentiate `f`.
+    delta: Size of the fundamental perturbation in the stencil.
+
+  Returns:
+    The approximate gradient.  Has the same structure as the return from a
+      corresponding call to `tf.GradientTape().gradient`.
+  """
+
+  def var_gradient(var):
```

Comment:
can you use var_jacobian for both of these approximate fns?

Reply:
They are different computations: this one sums over all entries of `f()`, corresponding to the behavior of `tape.gradient`, while `var_jacobian` does not perform such a sum

Reply:
Ah I thought you meant rename `var_gradient` to `var_jacobian`, but maybe you mean try to use `var_jacobian` in the gradient computation

Reply:
Looks like we could, but it would mean doing an unnecessary `transpose` and `reshape` during the gradient computation

Reply:
Ah the other sticky point is that `var_jacobian` assumes `f()` is a tensor, while `var_gradient` allows `f()` to be an arbitrary nested structure of tensors.  In particular `tf.map_fn` in `var_jacobian` would fail, since `_five_point_stencil` returns the same structure as `f()`

Reply:
So my conclusion is that we cannot without significant rethinking of the code

Reply:
fair enough - indeed was hoping we could use `var_jacobian` call for gradient special case